src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -975,6 +1017,60 @@ private async ValueTask<int> ReadAsyncInternal<TReadAdapter>(TReadAdapter adapte
             }
         }
 
+        // This function tries to make sure buffer has at least minSize bytes available.
+        // If we have enough data, it returns synchronously. If not, it will try to read
+        // remaining bytes from given stream.
+        private ValueTask<int> FillHandshakeBufferAsync<TReadAdapter>(TReadAdapter adapter, int minSize)
+            where TReadAdapter : ISslReadAdapter
+        {
+            if (_handshakeBuffer.ActiveLength >= minSize)
+            {
+                return new ValueTask<int>(minSize);
+            }
+
+            int bytesNeeded = minSize - _handshakeBuffer.ActiveLength;
+            _handshakeBuffer.EnsureAvailableSpace(bytesNeeded);
+
+            while (_handshakeBuffer.ActiveLength < minSize)
+            {
+                ValueTask<int> t = adapter.ReadAsync(_handshakeBuffer.AvailableMemory);
+                if (!t.IsCompletedSuccessfully)
+                {
+                    return InternalFillHandshakeBufferAsync(adapter, t, minSize);
+                }
+                int bytesRead = t.Result;
+                if (bytesRead == 0)
+                {
+                    return new ValueTask<int>(0);
+                }
+
+                _handshakeBuffer.Commit(bytesRead);
+            }
+
+            return new ValueTask<int>(minSize);
+
+            async ValueTask<int> InternalFillHandshakeBufferAsync(TReadAdapter adap,  ValueTask<int> task, int minSize)
+            {
+                while (true)
+                {
+                    int bytesRead = await task.ConfigureAwait(false);
+                    if (bytesRead == 0)
+                    {
+                        throw new IOException(SR.net_io_eof);
```

Comment:
Why does this throw for EOF but the fast-path earlier successfully returns 0 for it?  That means there's a race condition, where it may return 0 or may throw for EOF depending on how quickly the EOF is received.